feat(RTLplay): add presence

pierrequiroul opened the pull request:
## Description 
<!-- A clear and detailed description of the changes, referencing issues if applicable -->
Add presence for RTLplay. It's a streaming service where you can find all the programs live and in replay of RTL TVI, Club RTL, Plug RTL, Bel RTL and Radio Contact channels. Those channels are part of one of the main television group in Belgium.
It also support the Radio Contact webradio on its own website

List of assets and screenshots : https://imgur.com/a/rtlplay-premid-assets-iC5CIKJ

- [Previous PR](https://github.com/PreMiD/Presences/pull/8380)
- Resolves #7282

## Acknowledgements
- [X] I read the [Presence Guidelines](https://github.com/PreMiD/Presences/blob/main/.github/CONTRIBUTING.md)
- [X] I linted the code by running `yarn format`
- [X] The PR title follows the repo's [commit conventions](https://github.com/PreMiD/Presences/blob/main/.github/COMMIT_CONVENTION.md)

## Screenshots
<details>
<summary> Proof showing the creation/modification is working as expected </summary>

(Animated Logo)
![https://i.imgur.com/ah3LaMD.gif](https://i.imgur.com/ah3LaMD.gif)
![https://i.imgur.com/J1VmLe8.png](https://i.imgur.com/J1VmLe8.png)
![https://i.imgur.com/TWZ1Ion.png](https://i.imgur.com/TWZ1Ion.png)
![https://i.imgur.com/yBkCT3i.png](https://i.imgur.com/yBkCT3i.png)
![https://i.imgur.com/HtXwTTU.png](https://i.imgur.com/HtXwTTU.png)
![https://i.imgur.com/2ErwUzH.png](https://i.imgur.com/2ErwUzH.png)
![https://i.imgur.com/ukGiQ60.png](https://i.imgur.com/ukGiQ60.png)
![https://i.imgur.com/yhJI2cA.png](https://i.imgur.com/yhJI2cA.png)
![https://i.imgur.com/quruELz.png](https://i.imgur.com/quruELz.png)
![https://i.imgur.com/WiHLsUW.png](https://i.imgur.com/WiHLsUW.png)
![https://i.imgur.com/duxWfne.png](https://i.imgur.com/duxWfne.png)



</details>

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+		return url;
+	}
+}
+
```
```suggestion
let strings: Awaited<ReturnType<typeof getStrings>>

```

pierrequiroul replied:
I had to remove your line in because I already had it in line 41 :)

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				presence.getSetting<boolean>("usePosterImage"),
+			]);
+
+	if (oldLang !== lang) {
```
```suggestion
	if (oldLang !== lang || !strings) {
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+		case pathname === "/" ||
+			(pathname.split("/")[1] === "rtlplay" &&
+				pathname.split("/").length <= 2): {
+			presenceData.details = (await strings).browsing;
```
Now u can change the await (strings). to strings.
```suggestion
			presenceData.details = strings.browsing;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			presenceData.details = (await strings).browsing;
+
+			presenceData.smallImageKey = Assets.Reading;
+			presenceData.smallImageText = (await strings).browsing;
```
```suggestion
			presenceData.smallImageText = strings.browsing;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			presenceData.smallImageText = (await strings).browsing;
+
+			if (!privacy) {
+				presenceData.state = (await strings).viewHome;
```
```suggestion
				presenceData.state = strings.viewHome;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+
+		(https://www.rtlplay.be/rtlplay/recherche) */
+		case pathname.split("/")[2] === "recherche": {
+			if (privacy) presenceData.details = (await strings).searchSomething;
```
```suggestion
			if (privacy) presenceData.details = strings.searchSomething;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						.querySelector("ol.search__results")
+						?.getAttribute("data-tracking")
+				).searchQuery
+					? (await strings).search
+					: (await strings).searchSomething;
```
```suggestion
					? strings.search
					: strings.searchSomething;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				if (buttons) {
+					presenceData.buttons = [
+						{
+							label: (await strings).buttonViewPage, // Need to be a general string
```
```suggestion
							label: strings.buttonViewPage, // Need to be a general string
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			}
+
+			presenceData.smallImageKey = Assets.Search;
+			presenceData.smallImageText = (await strings).search;
```
```suggestion
			presenceData.smallImageText = strings.search;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+
+		(https://www.rtlplay.be/rtlplay/ma-liste) */
+		case ["ma-liste"].includes(pathname.split("/")[2]): {
+			presenceData.details = (await strings).browsing;
+			presenceData.state = (await strings).viewAPage;
```
```suggestion
			presenceData.details = strings.browsing;
			presenceData.state = strings.viewAPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			presenceData.details = (await strings).browsing;
+			presenceData.state = (await strings).viewAPage;
+			if (!privacy) {
+				presenceData.state = (await strings).viewList;
```
```suggestion
				presenceData.state = strings.viewList;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				if (buttons) {
+					presenceData.buttons = [
+						{
+							label: (await strings).buttonViewPage, // Need to be a general string
```
```suggestion
							label: strings.buttonViewPage, // Need to be a general string
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+		case ["collection", "series", "films", "divertissement"].includes(
+			pathname.split("/")[2]
+		): {
+			if (privacy) presenceData.details = (await strings).viewAPage;
```
```suggestion
			if (privacy) presenceData.details = strings.viewAPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						.textContent
+				);
+
+				presenceData.details = (await strings).viewCategory;
```
```suggestion
				presenceData.details = strings.viewCategory;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						: null;
+
+				presenceData.smallImageKey = Assets.Viewing;
+				presenceData.smallImageText = (await strings).viewCategory;
```
```suggestion
				presenceData.smallImageText = strings.viewCategory;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				if (buttons) {
+					presenceData.buttons = [
+						{
+							label: (await strings).buttonViewPage, // Need to be a general string
```
```suggestion
							label: strings.buttonViewPage, // Need to be a general string
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+					if (exist("i.playerui__icon--name-play")) {
+						// State paused
+						presenceData.smallImageKey = Assets.Pause;
+						presenceData.smallImageText = (await strings).pause;
```
```suggestion
						presenceData.smallImageText = strings.pause;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+					} else {
+						// State live
+						presenceData.smallImageKey = Assets.LiveAnimated;
+						presenceData.smallImageText = (await strings).live;
```
```suggestion
						presenceData.smallImageText = strings.live;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+					}
+
+					if (privacy) {
+						presenceData.details = (await strings).watchingLive;
```
```suggestion
						presenceData.details = strings.watchingLive;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+
+						presenceData.type = getChannel(pathname.split("/")[3]).type;
+
+						presenceData.details = (await strings).watchingLive;
```
```suggestion
						presenceData.details = strings.watchingLive;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						presenceData.type = getChannel(pathname.split("/")[3]).type;
+
+						presenceData.details = (await strings).watchingLive;
+						presenceData.state = document.querySelectorAll(
+							"li[aria-current='true'] > a > div > h2"
+						)[0].textContent;
```
```suggestion
						presenceData.state = document.querySelector(
							"li[aria-current='true'] > a > div > h2"
						).textContent;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						if (buttons) {
+							presenceData.buttons = [
+								{
+									label: (await strings).buttonWatchStream,
```
```suggestion
									label: strings.buttonWatchStream,
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						if (buttons) {
+							presenceData.buttons = [
+								{
+									label: (await strings).buttonListenAlong,
```
```suggestion
									label: strings.buttonListenAlong,
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			).groups || {};
+			if (privacy) {
+				presenceData.details =
+					episodeName !== ""
```
```suggestion
					!episodeName
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			if (privacy) {
+				presenceData.details =
+					episodeName !== ""
+						? (await strings).watchingShow
+						: (await strings).watchingMovie;
```
```suggestion
						? strings.watchingShow
						: strings.watchingMovie;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				if (usePresenceName) presenceData.name = mediaName;
+
+				presenceData.details =
+					episodeName !== ""
+						? `${(await strings).watching} ${mediaName}`
+						: (await strings).watchingMovie;
```
```suggestion
					!episodeName
						? `${strings.watching} ${mediaName}`
						: strings.watchingMovie;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				}
+
+				if (seasonNumber && episodeNumber) {
+					presenceData.largeImageText = `${
+						(await strings).season
+					} ${seasonNumber} - ${(await strings).episode} ${episodeNumber}`;
```
```suggestion
					presenceData.largeImageText = `${
						strings.season
					} ${seasonNumber} - ${strings.episode} ${episodeNumber}`;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+							label:
+								episodeName !== ""
+									? strings.buttonWatchEpisode
+									: (await strings).buttonWatchMovie, // Need to be a general string
```
```suggestion
									: strings.buttonWatchMovie, // Need to be a general string
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			if (exist("i.playerui__icon--name-play")) {
+				// State paused
+				presenceData.smallImageKey = Assets.Pause;
+				presenceData.smallImageText = (await strings).pause;
```
```suggestion
				presenceData.smallImageText = strings.pause;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				presenceData.smallImageText = (await strings).pause;
+			} else {
+				presenceData.smallImageKey = Assets.Play;
+				presenceData.smallImageText = (await strings).play;
```
```suggestion
				presenceData.smallImageText = strings.play;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+		(https://www.rtlplay.be/rtlplay/salvation~2ab30366-51fe-4b29-a720-5e41c9bd6991) */
+		case pathname.split("/")[2].length > 15: {
+			presenceData.smallImageKey = Assets.Viewing;
+			presenceData.smallImageText = (await strings).viewAPage;
```
```suggestion
			presenceData.smallImageText = strings.viewAPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			presenceData.smallImageKey = Assets.Viewing;
+			presenceData.smallImageText = (await strings).viewAPage;
+
+			if (privacy) presenceData.details = (await strings).viewAPage;
```
```suggestion
			if (privacy) presenceData.details = strings.viewAPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+					}`;
+				}
+
+				presenceData.details = (await strings).viewPage;
```
```suggestion
				presenceData.details = strings.viewPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			break;
+		}
+		default: {
+			presenceData.details = (await strings).viewAPage;
```
```suggestion
			presenceData.details = strings.viewAPage;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+
+				presenceData.details = (await strings).viewPage;
+				presenceData.state =
+					document.querySelectorAll("h1.detail__title")[0].textContent;
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
					document.querySelector("h1.detail__title").textContent;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				if (poster) {
+					presenceData.largeImageKey = await getShortURL(
+						document
+							.querySelectorAll("img.detail__poster")[0]
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
							.querySelector("img.detail__poster")
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+						if (
+							!useChannelName &&
+							document
+								.querySelectorAll("li[aria-current='true'] > a > div > h2")[0]
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
								.querySelector("li[aria-current='true'] > a > div > h2")
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+								.querySelectorAll("li[aria-current='true'] > a > div > h2")[0]
+								.textContent.toLowerCase() !== "aucune donnée disponible"
+						) {
+							presenceData.name = document.querySelectorAll(
+								"li[aria-current='true'] > a > div > h2"
+							)[0].textContent;
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
							presenceData.name = document.querySelector(
								"li[aria-current='true'] > a > div > h2"
							).textContent;
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+		(https://www.rtlplay.be/rtlplay/player/75e9a91b-29d1-4856-be8c-0b3532862404) */
+		case pathname.split("/")[2] === "player": {
+			const {
+				mediaName = document.querySelectorAll("h1.lfvp-player__title")[0]
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
				mediaName = document.querySelector("h1.lfvp-player__title")
```

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+				episodeNumber = "",
+				episodeName = "",
+			} = (
+				document
+					.querySelectorAll("h1.lfvp-player__title")[0]
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
				document
					.querySelector("h1.lfvp-player__title")
```

DarkVillager reviewed (changes requested):
Hey thanks for the PR.

If you have questions about the requested changes feel free to dm me on discord (@dark_ville)

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,555 @@
+			if (privacy) presenceData.details = (await strings).viewAPage;
+			else {
+				const data = JSON.parse(
+					document.querySelectorAll("script[type='application/ld+json']")[0]
+						.textContent
```
If you're getting the first element user querySelector() instead of querySelectorAll()
```suggestion
					document.querySelector("script[type='application/ld+json']")
						.textContent
```

DarkVillager reviewed (changes requested):
I forgot one...

pierrequiroul commented:
> Hey thanks for the PR.
> 
> If you have questions about the requested changes feel free to dm me on discord (@dark_ville)
> 
> _Yes i requested 40 changes, please add them to a batch and dont make them 40 commits_

Thanks for the feedbacks, I applied your change and I look into the code if something new needs to be corrected based on your feedback :)

DarkVillager commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,556 @@
+		return url;
+	}
+}
+let strings: Awaited<ReturnType<typeof getStrings>>
```
```suggestion
```

pierrequiroul replied:
Np, I saw that and I already remove it in #[6a5ac51](https://github.com/PreMiD/Presences/pull/8605/commits/6a5ac51197d512560dbf6a53883de90664ed412a) :)

DarkVillager reviewed (changes requested):
Nvm this was already at a different line.

pierrequiroul commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,547 @@
+	Logo = "https://i.imgur.com/KNsI47l.png",
+	Animated = "https://imgur.com/uAqZdFg.gif",
+	Deferred = "https://i.imgur.com/o69hUKD.gif",
+	LiveAnimated = "https://i.imgur.com/6oezfP6.gif",
```
Yes sorry, I forgot that guideline when I added those new assets :)
I fixed it in [#cdac88d](https://github.com/PreMiD/Presences/pull/8605/commits/cdac88de26b17d14a4c886a3944feb50efc0ad3f)

Bas950 commented on websites/R/RTLplay/presence.ts:
```diff
@@ -0,0 +1,547 @@
+	return document.querySelector(selector) !== null;
+}
+
+const shortenedURLs: Record<string, string> = {};
+async function getShortURL(url: string) {
+	if (!url || url.length < 256) return url;
+	if (shortenedURLs[url]) return shortenedURLs[url];
+	try {
+		const pdURL = await (
+			await fetch(`https://pd.premid.app/create/${url}`)
+		).text();
+		shortenedURLs[url] = pdURL;
+		return pdURL;
+	} catch (err) {
+		presence.error(err);
+		return url;
+	}
+}
```
This is done automatically in the extension

DarkVillager replied:
wait it is?

pierrequiroul replied:
I removed these lines in [#2a79bbc](https://github.com/PreMiD/Presences/pull/8605/commits/2a79bbc03c35f1a01c696a8cd087bc9b4d16a672)
I also added an additionnal strings to translate Deferred with a function, idk if there's a better option at my level and I also tweaked little things :)

Bas950 commented on websites/R/RTLplay/presence.ts:
```diff
@@ -105,20 +129,33 @@ function exist(selector: string) {
 	}
+
+	return Math.round(totalSeconds / 60);
 }
```
I think there is a function for this in the presence class? Take a look at `presence.timestampFromFormat`

pierrequiroul replied:
Thanks, I've just tried it. I hadn't realized that it gave a format in seconds, I can convert it to minutes like the initial function. :)

pierrequiroul commented:
[#b68a248](https://github.com/PreMiD/Presences/pull/8605/commits/b68a248cf9a57f8a3583bb58e2092be1a18ec48a)
With Bas950's changes, I've also added the music currently playing on the radio. Since the livestream is the video version of what's being played on the radio, I can retrieve the information via the official api used on web radio sites. I've added a localized asset for ads in English and French when there's a commercial. And I checked again if the presence works in different case scenarios. :)
![Radiolive](https://github.com/user-attachments/assets/84220444-43b9-48b8-bfcd-06b2c97410d4)
![nosong](https://github.com/user-attachments/assets/6cbaa608-124f-4045-825b-d21b7c3158f0)
![Ad2](https://github.com/user-attachments/assets/5fc09321-cda2-4e6e-b8a6-31b24d627a0f)
![Ad](https://github.com/user-attachments/assets/015993ed-b856-4dbe-87c2-f75794c6a708)

pierrequiroul commented:
Hi it's me again. I added a function to generate a cool image uncropped if it's landscape or portrait mode.
![Landscape](https://github.com/user-attachments/assets/66fda9b7-63f6-4d67-b021-35bfcaed9fb1)
![Portrait](https://github.com/user-attachments/assets/c98714f5-1c54-41b2-9b2c-41028835fb37)
